### PR TITLE
Add `dropout_path` as a structured form of `dropout_edge`

### PR DESCRIPTION
**Pull request by EdisonLeeeee**
This PR implements `DropPath` from [MaskGAE: Masked Graph Modeling Meets Graph Autoencoders](https://arxiv.org/abs/2205.10053). DropPath is a structured form of DropEdge, which drops a group of edges (paths) based on random walks.

DropPath follows three steps to sample edges to drop:
+ Sample a set of root nodes $R$ with probability `r` from a Bernoulli distribution. (0<=`r`<=1), 
+ Perform random walks starting from root nodes $R$
+ Drop edges sampled by random walks

Link to https://github.com/pyg-team/pytorch_geometric/issues/5452

### CODE REVIEW

File: torch_geometric/utils/dropout.py [hunk 1]
```diff
@@ -208,3 +214,94 @@ def dropout_edge(edge_index: Tensor, p: float = 0.5,
...
+        edge_index (LongTensor): The edge indices.
+        r (float, optional): The ratio of source nodes to start
+            random walks from. (default: :obj:`0.5`)
+        force_undirected (bool, optional): If set to :obj:`True`, will either
+            drop or keep both edges of an undirected edge.
+            (default: :obj:`False`)
+        walks_per_node (int, optional): The number of walks per node.
+            (default: :obj:`1`)
+        walk_length (int, optional): The length per walk. (default: :obj:`3`)
+        p (float, optional): :obj:`p` in random walks. (default: :obj:`1`)
```
Comment (rusty1s) [inline torch_geometric/utils/dropout.py:242]: Let's link to `Node2Vec` implementation for `p` and `q` meaning?

File: torch_geometric/utils/dropout.py [hunk 2]
```diff
@@ -208,3 +214,94 @@ def dropout_edge(edge_index: Tensor, p: float = 0.5,
...
+    if not training or r == 0.0:
+        return edge_index, edge_mask
+
+    if random_walk is None:
+        raise ImportError('`dropout_path` requires `torch-cluster`.')
+
+    row, col = edge_index
+    num_nodes = maybe_num_nodes(edge_index, num_nodes)
+    deg = torch.zeros(num_nodes, device=row.device)
+    deg.scatter_add_(0, row, torch.ones(row.size(0), device=row.device))
```
Comment (rusty1s) [inline torch_geometric/utils/dropout.py:285]: Let's use `utils.degree` for this.

File: torch_geometric/utils/dropout.py [hunk 3]
```diff
@@ -208,3 +214,94 @@ def dropout_edge(edge_index: Tensor, p: float = 0.5,
...
+
+    num_starts = round(r * num_nodes)
+    start = torch.randperm(num_nodes, device=edge_index.device)[:num_starts]
+    start = start.repeat(walks_per_node)
+
+    rowptr = row.new_zeros(num_nodes + 1)
```
Comment (rusty1s) [inline torch_geometric/utils/dropout.py:291]: This does not look correct to me. We would first need to sort `edge_index` row-wise. Otherwise, `col` does not align with `rowptr`. Note that in this case, you also need to permute the resulting `edge_id`.

Comment (EdisonLeeeee) [reply torch_geometric/utils/dropout.py:299]: Thanks! It has been fixed with an `is_sorted` argument.

**Comment by EdisonLeeeee** [comment]
A quick question (maybe stupid): how to disable the GitHub bot that always removes labels when I push the code? That's a bit annoying :-(

File: torch_geometric/utils/dropout.py (hunk 3 above)
Comment (EdisonLeeeee) [reply torch_geometric/utils/dropout.py:299]: Not yet. Also need to permute the resulting `edge_id`.

Comment (EdisonLeeeee) [reply torch_geometric/utils/dropout.py:292]: Done. Argument `force_undirected` has been dropped to make it simple.

**Review by wsad1 — APPROVED** [review]
Thanks for adding this.

File: test/utils/test_dropout.py [hunk 4]
```diff
@@ -59,3 +64,33 @@ def test_dropout_edge():
     out = dropout_edge(edge_index, force_undirected=True)
     assert out[0].tolist() == [[0, 1, 1, 2], [1, 2, 0, 1]]
     assert out[1].tolist() == [0, 2, 0, 2]
+
+
+def test_dropout_path():
```
Comment (wsad1) [inline test/utils/test_dropout.py:69]: Run this test only when `torch_cluster` is available.
`@withPackage('torch_cluster')`

File: torch_geometric/utils/dropout.py [hunk 5]
```diff
@@ -208,3 +216,88 @@ def dropout_edge(edge_index: Tensor, p: float = 0.5,
...
+    r"""Drops edges from the adjacency matrix :obj:`edge_index`
+    based on random walks. The source nodes to start random walks from are
+    sampled with probability :obj:`r` from a Bernoulli distribution.
+
+    The method returns (1) the retained :obj:`edge_index`, (2) the edge mask
+    indicating which edges were retained.
+
+    Args:
+        edge_index (LongTensor): The edge indices.
+        r (float, optional): The ratio of source nodes to start
```
Comment (rusty1s) [inline torch_geometric/utils/dropout.py:234]: A bit sad that we can not call this `p`. How important are the `node2vec` parameters here in your opinion?

File: torch_geometric/utils/dropout.py [hunk 6]
```diff
@@ -208,3 +216,88 @@ def dropout_edge(edge_index: Tensor, p: float = 0.5,
...
+    if not is_sorted:
+        edge_orders = torch.arange(num_edges, device=edge_index.device)
+        edge_index, edge_orders = sort_edge_index(edge_index, edge_orders,
+                                                  num_nodes=num_nodes)
+
+    row, col = edge_index
+    deg = degree(row, num_nodes=num_nodes)
+    num_starts = round(r * num_nodes)
+    start = torch.randperm(num_nodes, device=edge_index.device)[:num_starts]
+    start = start.repeat(walks_per_node)
```
Comment (rusty1s) [inline torch_geometric/utils/dropout.py:290]: This is a bit different from the usual definition of
```python
edge_mask = torch.rand(row.size(0), device=edge_index.device) >= r
start = edge_mask.nonzero().view(-1)
```
Shall we align this?

**Review by rusty1s — APPROVED** [review]

**Comment by rusty1s** [comment]
Let's also add it to the `CHANGELOG.md`. Feel free to merge afterwards.

**Comment by EdisonLeeeee** [comment]
Since there are some big changes made, I'm afraid it is still necessary to check it again :(

File: torch_geometric/utils/dropout.py [hunk 7]
```diff
@@ -208,3 +216,82 @@ def dropout_edge(edge_index: Tensor, p: float = 0.5,
...
+    edge_orders = None
+    ori_edge_index = edge_index
+    if not is_sorted:
+        edge_orders = torch.arange(num_edges, device=edge_index.device)
+        edge_index, edge_orders = sort_edge_index(edge_index, edge_orders,
+                                                  num_nodes=num_nodes)
+
+    row, col = edge_index
+    sample_mask = torch.rand(row.size(0), device=edge_index.device) <= p
+    start = row[sample_mask].repeat(walks_per_node)
```
Comment (wsad1) [inline torch_geometric/utils/dropout.py:283]: Doesn't this mean we'll start many more than `walks_per_node` random walks for nodes with a high out-degree.
Sorry for the back and forth, but I felt the way you defined start perviously was better and actually meant we start exactly `walks_per_node` random walks`.

Comment (EdisonLeeeee) [reply torch_geometric/utils/dropout.py:283]: Agreed. Or maybe we can add an option such as `walk_by='edge'` or `walk_by='node'` to align them?

Comment (wsad1) [reply torch_geometric/utils/dropout.py:283]: Maybe I am missing something. Why do we want to support `walk_by='edge'`?

Comment (EdisonLeeeee) [reply torch_geometric/utils/dropout.py:283]: Oh sorry I meant we can sample starting nodes node-wise (previous version) or edge-wise (current version).  For edge-wise sampling, as you mentioned, we can start random walks for nodes with a high out-degree, this seems to make sense for graphs with particular imbalanced structures.

File: torch_geometric/utils/dropout.py [hunk 8]
```diff
@@ -208,3 +216,82 @@ def dropout_edge(edge_index: Tensor, p: float = 0.5,
...
+        edge_index (LongTensor): The edge indices.
+        p (float, optional): Sample probability. (default: :obj:`0.2`)
+        walks_per_node (int, optional): The number of walks per node, same as
+            :class:`~torch_geometric.nn.models.Node2Vec`. (default: :obj:`1`)
+        walk_length (int, optional): The walk length, same as
+            :class:`~torch_geometric.nn.models.Node2Vec`. (default: :obj:`3`)
+        num_nodes (int, optional): The number of nodes, *i.e.*
+            :obj:`max_val + 1` of :attr:`edge_index`. (default: :obj:`None`)
+        is_sorted (bool, optional): If set to :obj:`True`, will expect
+            :obj:`edge_index` to be already sorted row-wise.
```
Comment (rusty1s) [inline torch_geometric/utils/dropout.py:243]:
```suggestion
            :obj:`edge_index` to be already sorted row-wise. (default: :obj:`False`)
```

**Review by rusty1s — APPROVED** [review]
Looks perfect :)

File: torch_geometric/utils/dropout.py (hunk 7 above)
Comment (EdisonLeeeee) [reply torch_geometric/utils/dropout.py:283]: I think this is the last one that needs to be resolved.

Comment (rusty1s) [reply torch_geometric/utils/dropout.py:283]: No strong opinion. I think we can safely merge this and wait on user feedback. To me, it makes sense to sample per edge since we are dropping paths.

Comment (rusty1s) [reply torch_geometric/utils/dropout.py:283]: Maybe just add a comment for now?

File: torch_geometric/utils/dropout.py [hunk 9]
```diff
@@ -208,3 +216,82 @@ def dropout_edge(edge_index: Tensor, p: float = 0.5,
...
+        edge_index (LongTensor): The edge indices.
+        p (float, optional): Sample probability. (default: :obj:`0.2`)
+        walks_per_node (int, optional): The number of walks per node, same as
+            :class:`~torch_geometric.nn.models.Node2Vec`. (default: :obj:`1`)
+        walk_length (int, optional): The walk length, same as
+            :class:`~torch_geometric.nn.models.Node2Vec`. (default: :obj:`3`)
+        num_nodes (int, optional): The number of nodes, *i.e.*
+            :obj:`max_val + 1` of :attr:`edge_index`. (default: :obj:`None`)
+        is_sorted (bool, optional): If set to :obj:`True`, will expect
+            :obj:`edge_index` to be already sorted row-wise. (default: :obj:`False`)
```
Comment (rusty1s) [inline torch_geometric/utils/dropout.py:243]:
```suggestion
            :obj:`edge_index` to be already sorted row-wise.
            (default: :obj:`False`)
```

File: torch_geometric/utils/dropout.py (hunk 7 above)
Comment (EdisonLeeeee) [reply torch_geometric/utils/dropout.py:284]: > Maybe just add a comment for now and clarify in the doc?
Good idea.

Comment (EdisonLeeeee) [reply torch_geometric/utils/dropout.py:284]: Alternatively, we can let `p` accept `Tensor` to specify the nodes to start random walks, rather than randomly sampled from nodes and edges. Let's leave it as future work and merge it now.

Comment (wsad1) [reply torch_geometric/utils/dropout.py:284]: Apologies, i forgot to respond. its good that you merged it :).

Comment (EdisonLeeeee) [reply torch_geometric/utils/dropout.py:284]: No worries :)
